extensions/azure/cosmos/cosmos-common/src/main/java/org/eclipse/dataspaceconnector/azure/cosmos/CosmosDbApiImpl.java

```diff
@@ -31,26 +29,21 @@
 import com.azure.cosmos.models.PartitionKey;
 import com.azure.cosmos.models.SqlParameter;
 import com.azure.cosmos.models.SqlQuerySpec;
-import org.eclipse.dataspaceconnector.common.string.StringUtils;
 import org.eclipse.dataspaceconnector.spi.EdcException;
 import org.eclipse.dataspaceconnector.spi.persistence.EdcPersistenceException;
-import org.eclipse.dataspaceconnector.spi.security.Vault;
 import org.eclipse.dataspaceconnector.spi.system.health.HealthCheckResult;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.List;
 import java.util.Scanner;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public class CosmosDbApiImpl implements CosmosDbApi {
```

Comment:
Wouldn't it also have been possible to create a subclass for the `CosmosDbApiImpl`, e.g. named `TestCosmosDbApiImpl` that simply overrides the `createClient` method (which doesn't need to be static anyways) and uses the `CosmosTestClient`? That way we would be able to contain the `CosmosClient` (which is a third-party class) and not expose it.

Reply:
To me creating a client is a thing that goes over the 'CosmosDbApi' component responsibilities. 
`CosmosClient` is a third party class indeed, but a `CosmosDbApi` class could not exist without it, so exposing it is not a big issue.

Reply:
Well, in this case it would have saved the creation of another extension, plus another "factory" interface + impl.

Reply:
the extension model is a base concept of the EDC, I don't see nothing wrong in using it, in this case it will permit users to customize the `CosmosClient` at extension level, like, for example, to accomplish specific needs like using a different authentication type or adding more than one preferred regions.

Making the `Api` instantiate the client, other than the responsibility issue I mentioned in the previous comment,  would need to expose a `protected getClient` method that would be needed only for testing purposes, that's an habit to be avoided if possible.

Reply:
The missing piece in my mind was the fact that when using the `CosmosClient` for testing with the CosmosDB emulator, it requires some special setup (downloading and storing a certificate), in connection with the fact that that we're dealing with and e2e test, which uses EDC's dependency injection capabilities. That means, that the "local" client must be supplied through DI.

Just one caveat: the CosmosDB emulator is not suitable for usage with CI/CD pipelines, it simply is not stable enough. Thus, E2E tests should run against a real cosmos instance, or not run at all, if no `COSMOS_KEY` is present. Same way we do for the other cosmos integration tests.